lib/lower/Runtime.cpp

```diff
@@ -68,91 +70,37 @@ RuntimeStaticDescription::getFuncOp(mlir::OpBuilder &builder) const {
   }
   return function;
 }
-} // namespace Fortran::lower
-
-// TODO remove dependencies to stub rt below
-
-namespace Br = Fortran::lower;
-using namespace Fortran;
-using namespace Fortran::lower;
-
-namespace {
-
-using FuncPointer = llvm::SmallVector<mlir::Type, 4> (*)(mlir::MLIRContext *,
-                                                         int kind);
-
-// FIXME: these are wrong and just serving as temporary placeholders
-using Cplx = mlir::IntegerType;
-using Intopt = mlir::IntegerType;
-using TypeDesc = mlir::IntegerType;
-
-template <typename A, typename... B>
-void consType(llvm::SmallVector<mlir::Type, 4> &types, mlir::MLIRContext *ctx,
-              int kind) {
-  if constexpr (std::is_same_v<A, mlir::IntegerType>) {
-    types.emplace_back(A::get(kind, ctx));
-  } else if constexpr (std::is_same_v<A, mlir::ComplexType>) {
-    // FIXME: this is wrong
-    assert(false);
-    types.emplace_back(mlir::IntegerType::get(kind, ctx));
-  } else {
-    types.emplace_back(A::get(ctx, kind));
-  }
-  if constexpr (sizeof...(B) > 0) {
-    consType<B...>(types, ctx, kind);
-  }
-}
 
-template <typename... A>
-llvm::SmallVector<mlir::Type, 4> consType(mlir::MLIRContext *ctx, int kind) {
-  llvm::SmallVector<mlir::Type, 4> types;
-  if constexpr (sizeof...(A) > 0) {
-    consType<A...>(types, ctx, kind);
-  } else {
-    (void)ctx;
-    (void)kind;
-  }
-  return types;
-}
-
-#define DEFINE_RUNTIME_ENTRY(A, B, C, D) "Fortran_" B,
-char const *const RuntimeEntryNames[FIRT_LAST_ENTRY_CODE] = {
-#include "runtime.def"
+class RuntimeEntryDescription : public RuntimeStaticDescription {
+public:
+  using Key = RuntimeEntryCode;
+  constexpr RuntimeEntryDescription(Key k, const char *s, MaybeTypeCode r,
+                                    TypeCodeVector a)
+      : RuntimeStaticDescription{s, r, a}, key{k} {}
+  Key key;
 };
-
-#define UNPACK_TYPES(...) __VA_ARGS__
-#define DEFINE_RUNTIME_ENTRY(A, B, C, D) consType<UNPACK_TYPES C>,
-FuncPointer RuntimeEntryInputType[FIRT_LAST_ENTRY_CODE] = {
-#include "runtime.def"
+using RT = RuntimeStaticDescription;
+using RType = typename RT::TypeCode;
+using Args = typename RT::TypeCodeVector;
+using RTC = RuntimeEntryCode;
+static constexpr RuntimeEntryDescription runtimeTable[]{
+    {RTC::StopStatement, "StopStatement", RT::voidTy,
+     Args::create<RType::i32, RType::boolean, RType::boolean>()},
+    {RTC::StopStatementText, "StopStatementText", RT::voidTy,
+     Args::create<RType::charPtr, RType::i32, RType::boolean,
+                  RType::boolean>()},
+    {RTC::FailImageStatement, "StopStatementText", RT::voidTy,
+     Args::create<>()},
```

Comment:
What I think would be nice is to have a tablegen defn for this stuff. It would let us build the table(s) and factor out the redundancies easily.